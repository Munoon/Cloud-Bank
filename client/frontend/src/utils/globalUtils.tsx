import React from 'react';
import ReactDOM from 'react-dom';
import {Store} from "redux";
import {Provider} from "react-redux";

import applicationSettings from "./settings";

export function render(component: React.ReactElement, store?: Store) {
    if (applicationSettings.testEnvironment) {
        return;
    }

    let renderComponent = store ? (
        <Provider store={store}>
            {component}
        </Provider>
    ) : component;

    if (applicationSettings.development) {
        renderComponent = (
            <React.StrictMode>
                {renderComponent}
            </React.StrictMode>
        );
    }

    ReactDOM.render(renderComponent, document.getElementById('root'));
}

export function readProperty(key: string): string {
    if (applicationSettings.useCustomProperties) {
        return applicationSettings.customProperties[key];
    }

    let element = document.querySelector<HTMLMetaElement>(`meta[name="${key}"]`);
    return element ? element.content : null;
}

type TestIdType = { 'data-testid': string };

export function testId(key: string): TestIdType {
    if (!applicationSettings.testEnvironment) {
        return null;
    }
    return { "data-testid": key };
}