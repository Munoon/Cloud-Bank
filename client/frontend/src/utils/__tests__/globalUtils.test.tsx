import React from "react";

import {readProperty, render} from "../globalUtils";
import mockStore from "../../page/login/__tests__/utils/mockStore";
import {render as reactRender} from 'react-dom';

jest.mock('react-dom', () => ({
    render: jest.fn()
}));

jest.mock('../settings', () => ({
    useCustomProperties: false,
    testEnvironment: false,
    development: false
}));

describe('render test', () => {
    beforeEach(() => jest.resetAllMocks());

    it('Render test without store', () => {
        render(<h1>test</h1>);
        expect(reactRender).toHaveBeenCalledTimes(1);
        // @ts-ignore
        expect(reactRender.mock.calls[0]).toMatchSnapshot();
    });

    it('Render test with store', () => {
        render(<h1>test</h1>, mockStore());
        expect(reactRender).toHaveBeenCalledTimes(1);
        // @ts-ignore
        expect(reactRender.mock.calls[0]).toMatchSnapshot();
    });
});

describe('read property test', () => {
    let querySelectorSpy: jest.Mock<any, any[]>;
    let querySelectorProxy = (selector: string) => querySelectorSpy(selector);

    beforeAll(() => {
        Object.defineProperty(global.document, 'querySelector', { value: querySelectorProxy });
    });

    it('read key test', () => {
        querySelectorSpy = jest.fn(() => ({
            content: 'test'
        }));

        const result = readProperty('key');
        expect(result).toEqual('test');

        const calls: any[] = querySelectorSpy.mock.calls;
        expect(calls[calls.length - 1][0]).toEqual('meta[name="key"]');
    });

    it('read key, that not exist test', () => {
        querySelectorSpy = jest.fn(() => null);

        const result = readProperty('key');
        expect(result).toEqual(null);
    });
});