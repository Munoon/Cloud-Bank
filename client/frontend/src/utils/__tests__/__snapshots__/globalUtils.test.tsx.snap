// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render test Render test with store 1`] = `
Array [
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <h1>
      test
    </h1>
  </Provider>,
  null,
]
`;

exports[`render test Render test without store 1`] = `
Array [
  <h1>
    test
  </h1>,
  null,
]
`;
